smapi_test.go

```diff
@@ -559,6 +561,212 @@ func TestClientSave(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestCreateToken(t *testing.T) {
+	orgs := orgs()
+	testTenant := orgs.findTenantByOrg(1000)
+
+	url, mux, cleanup := newTestServer(t)
+	defer cleanup()
+
+	var called bool
+
+	mux.Handle("/api/v1/token/create", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		called = true
+
+		if err := requireMethod(w, r, http.MethodPost); err != nil {
+			return
+		}
+
+		if _, err := requireAuth(orgs, w, r, testTenant.id); err != nil {
+			return
+		}
+
+		writeResponse(
+			w,
+			http.StatusOK,
+			&model.TokenCreateResponse{
+				Msg:         "a message",
+				AccessToken: "a token",
+			},
+		)
+	}))
+
+	t.Run("valid token", func(t *testing.T) {
+		c := NewClient(url, testTenant.token, http.DefaultClient)
+		ctx := context.Background()
+		called = false
+		newToken, err := c.CreateToken(ctx)
+		require.True(t, called)
+		require.NoError(t, err)
+		require.NotEmpty(t, newToken)
+	})
+
+	t.Run("valid token", func(t *testing.T) {
+		token := invalidToken
+		c := NewClient(url, token, http.DefaultClient)
+		ctx := context.Background()
+		called = false
+		newToken, err := c.CreateToken(ctx)
+		require.True(t, called)
+		require.Error(t, err)
+		require.Empty(t, newToken)
+	})
+}
+
+func TestDeleteToken(t *testing.T) {
+	orgs := orgs()
+	testTenant := orgs.findTenantByOrg(1000)
+
+	url, mux, cleanup := newTestServer(t)
+	defer cleanup()
+
+	var called bool
+
+	mux.Handle("/api/v1/token/delete", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		called = true
+
+		if err := requireMethod(w, r, http.MethodDelete); err != nil {
+			return
+		}
+
+		if _, err := requireAuth(orgs, w, r, testTenant.id); err != nil {
+			return
+		}
+
+		writeResponse(
+			w,
+			http.StatusOK,
+			&model.TokenDeleteResponse{
+				Msg: "a message",
+			},
+		)
+	}))
+
+	t.Run("valid token", func(t *testing.T) {
+		c := NewClient(url, testTenant.token, http.DefaultClient)
+		ctx := context.Background()
+		called = false
+		err := c.DeleteToken(ctx)
+		require.True(t, called)
+		require.NoError(t, err)
+		require.Empty(t, c.accessToken)
+	})
+
+	t.Run("invalid token", func(t *testing.T) {
+		token := invalidToken
+		c := NewClient(url, token, http.DefaultClient)
+		ctx := context.Background()
+		called = false
+		err := c.DeleteToken(ctx)
+		require.True(t, called)
+		require.Error(t, err)
+		require.Equal(t, token, c.accessToken)
+	})
+}
+
+func TestRefreshToken(t *testing.T) {
+	orgs := orgs()
+	testTenant := orgs.findTenantByOrg(1000)
+	newToken := "a new token"
+
+	url, mux, cleanup := newTestServer(t)
+	defer cleanup()
+
+	var called bool
```

Comment:
Currently in these tests we repeat `called = false` a lot, in each of the `t.Run` blocks. Personally I think it would be simpler and easier to understand if you just initialise this to `false` here instead.
```suggestion
	called := false
```

Reply:
While it's the same (bool's zero value is false) I avoided using short initialization precisely to avoid giving the impression that the variable is correctly initialized before the test runs.

Note that the structure of the test is pretty nasty: a server runs in a separate goroutine (so that the test is not blocked) and the `called` variable is modified there every time the particular path is accessed. Then _another_ goroutine is created where the subtest runs and called is initialized before calling the function that will cause the handler function to run. For the next subtest, called is initialized to false again so that we know that if the value changes, it changes because the handler function ran again. And the only reason why this whole thing exists is because there was a test that was passing without ever accessing the corresponding endpoint. And then it would have been odd to have some tests keep track of this and others not.

Reply:
Ah nice, thanks. I had expected that the setup and teardown parts would each run multiple times - I forgot about goroutines.
Confirmed [here](https://blog.golang.org/subtests#:~:text=the%20setup%20and%20tear-down%20code%20will%20run%20if%20any%20of%20the%20enclosed%20subtests%20are%20run%20and%20will%20run%20at%20most%20once) (pasting link for my own benefit).

Reply:
I don't really remember which one.
 
Regarding a mock... maybe. The HTTP server provided by `httptest` is designed with this scenario in mind: you have a client that sends HTTP requests and you want to control and record the responses. It's probably written more with the "test a mux" case in mind. In the option you are talking about, the thing that you'd be mocking wouldn't be the _server_ but the client (smapi.Client receives an http.Client as argument, and the only method we use is `Do` so it's super easy to replace with an interface) where you can examine the request and produce whatever `http.Response` you want. I'm not sure it would change the structure of the test a lot, though. It would obviously eliminate the HTTP server but the thing that has to replace it wouldn't be entirely trivial.